File: src/web/magmac/app/compile/StagedCompiler.ts
import { CompileResult } from "../../../magmac/app/compile/error/CompileResult";
import { Serializable } from "../../../magmac/app/lang/Serializable";
import { JavaLang } from "../../../magmac/app/lang/java/JavaLang";
import { Generator } from "../../../magmac/app/stage/generate/Generator";
import { Parser } from "../../../magmac/app/stage/parse/Parser";
import { UnitSet } from "../../../magmac/app/stage/unit/UnitSet";
export class StagedCompiler<R extends Serializable> {
	 StagedCompiler( parser : Parser<JavaLang.JavaRoot, R>,  generator : Generator) : public {this.parser=parser;this.generator=generator;;}
	public parseAndGenerate( units : UnitSet<JavaLang.JavaRoot>) : CompileResult<UnitSet<String>> {return this.parser.apply( units).mapValue( 0).flatMapValue( this.generator.apply);;}
}

File: packages/compiler-java/src/main/java/magmac/app/compile/StagedCompiler.java
package magmac.app.compile;

import magmac.app.compile.error.CompileResult;
import magmac.app.lang.Serializable;
import magmac.app.lang.java.JavaLang;
import magmac.app.stage.generate.Generator;
import magmac.app.stage.parse.Parser;
import magmac.app.stage.unit.UnitSet;

public final class StagedCompiler<R extends Serializable> implements Compiler {
    private final Parser<JavaLang.Root, R> parser;
    private final Generator generator;

    public StagedCompiler(Parser<JavaLang.Root, R> parser, Generator generator) {
        this.parser = parser;
        this.generator = generator;
    }

    @Override
    public CompileResult<UnitSet<String>> parseAndGenerate(UnitSet<JavaLang.Root> units) {
        return this.parser.apply(units)
                .mapValue((UnitSet<R> roots) -> roots.mapValues(Serializable::serialize))
                .flatMapValue(this.generator::apply);
    }
}